### ISSUE TITLE
How to create dependent rule

### ISSUE THREAD

**Issue by GrantHoffman** [bchavez/Bogus#293]
This has most of the functionality I need, Im just wondering if it's possible to do the following:
Imagine I have a DTO (pseudo code)

    EffectivePayRange
    {
         Int Pay; 
         DateTime StartDate ;
         DateTime EndDate;
    }

Is there a way to create a fake which is dependent on the other generation 

I want to be assure the start date is generated before the end date and I want the next start Date to be generated a day after the previously generated end dates so the dates are gapless and chained together

   e.g I would like to do something like this:

            var faker = new Faker<DivisionPayCodeTable>(); 
            faker.AggregateRuleFor(x => x.StartDate, (f, previousObjectGenerated) => previousObjectGenerated.EndDate.AddDays(1))
           // Automatically, creates mapping order so that start date is always set before endDate
           fake.SelfDependentRule(x => x.EndDate, (f, objectInGeneration) => objectInGeneration.StartDate.AddDays(10))

**Comment by bchavez** [comment]
Hi Grant,

Thank you for your question. There are a few ways you could do this but the basic idea behind all this is to maintain state in a variable that stores the last generated object.

#### For the given POCO object:
```csharp
public class EffectivePayRange
{
   public Int32 Pay;
   public DateTime StartDate;
   public DateTime EndDate;
}
```

## 1. `Faker<T>` using a closure
Probably the simplest of them all.
```csharp
void Main()
{
   EffectivePayRange prev = null;
   
   var faker = new Faker<EffectivePayRange>()
       .RuleFor(x => x.StartDate, f => prev?.EndDate.AddDays(1) ?? DateTime.Parse("1/1/2020"))
       .RuleFor(x => x.EndDate, (f, current) => current.StartDate.AddDays(10))
       .RuleFor(x => x.Pay, f => f.Random.Number(20, 80))
       .FinishWith( (f, current) => prev = current );

   faker.Generate(20).Dump();
}
```

## 2. Derive from `Faker<T>`
Deriving from `Faker<T>` is perfectly acceptable too.
```csharp
void Main()
{ 
   var faker = new StatefulFaker<EffectivePayRange>()
       .StatefulRuleFor(x => x.StartDate, (f, prev) => prev?.EndDate.AddDays(1) ?? DateTime.Parse("1/1/2020"))
       .RuleFor(x => x.EndDate, (f, current) => current.StartDate.AddDays(10))
       .RuleFor(x => x.Pay, f => f.Random.Number(20, 80));
       
   faker.Generate(20).Dump();
}

public class StatefulFaker<T> : Faker<T> where T : class
{
   protected T Previous { get;set; }
   
   public StatefulFaker()
   {
      this.FinishWith((f, t) => this.Previous = t);
   }
   
   public StatefulFaker<T> StatefulRuleFor<TProperty>(Expression<Func<T, TProperty>> property, Func<Faker, T, TProperty> setter) 
   {
      this.RuleFor(property, (f, current) => setter(f, this.Previous));
      return this;
   }
}
```

## 3. Stateful Extension for `Faker<T>`
Little bit hackish but works.
```csharp
void Main()
{
   var faker = new Faker<EffectivePayRange>()
       .StatefulRuleFor(x => x.StartDate, (f, prev) => prev?.EndDate.AddDays(1) ?? DateTime.Parse("1/1/2020"))
       .RuleFor(x => x.EndDate, (f, current) => current.StartDate.AddDays(10))
       .RuleFor(x => x.Pay, f => f.Random.Number(20, 80));

   faker.Generate(20).Dump();
}

public static class StatefulExtensionsForFakerT
{
   private static object Previous { get; set; }

   public static Faker<T> StatefulRuleFor<T, TProperty>(this Faker<T> faker, Expression<Func<T, TProperty>> property, Func<Faker, T, TProperty> setter) where T : class
   {
      faker.RuleFor(property, (f, current) => setter(f, Previous as T));
      faker.FinishWith( (f, current) => Previous = current);
      return faker;
   }
}
```
-----

All three solutions here produce very similar results as shown below:

![image](https://user-images.githubusercontent.com/478118/80854104-4c7ba080-8bea-11ea-90f1-d30e488b1385.png)

I hope that helps! Please let me know if you have any other questions. Also, have a great weekend.

Thanks,
Brian

**Comment by GrantHoffman** [comment]
@bchavez Thanks,  I just had one more quick question.  Are the rules applied in the order they're placed in the config?

**Comment by bchavez** [comment]
Hi @GrantHoffman,

The TL;DR; is that rules are always applied in the way they are initially defined. See #279 for more details.

More concretely, consider the following code:
```csharp
public class User{
   public string FirstName;
   public char MiddleInitial;
   public string LastName;
}
```
```csharp
 1  void Main()
 2  {
 3     var userFaker = new Faker<User>()
 4     .RuleFor(u => u.FirstName, f => f.Name.FirstName())
 5     .RuleFor(u => u.MiddleInitial, (f, u) => u.FirstName[0])
 6     .RuleFor(u => u.LastName, f => f.Name.LastName());
 7     
 8     // works fine, no issue.
 9     userFaker.Generate().Dump();
10     
11     // Re-define the rule to be dependent on the LastName instead.
12     userFaker.RuleFor(u => u.MiddleInitial, (f, u) => u.LastName[0]);
13     // null reference exception is thrown below
14     userFaker.Generate().Dump();
15  }
```
In the code above, as initially defined, **Line 5**, the `u.MiddleInitial` character is dependent on the first name character of `u.FirstName`. 

On **Line 12**, we re-define the `u.MiddleInitial` character to be dependent on the `u.LastName`. However, a null reference exception is thrown on **Line 14** because **Line 12** is "replacing" the `u.MiddleInitial` rule slot with a new *dependent* rule on a property whose value has not been generated yet.

The fix, of course, is to move the `u.MiddleInitial` rule below `u.FirstName` and `u.LastName` so that `u.FirstName` and `u.LastName` are defined first. In other words, the following code does not throw a null reference exception when `u.MiddleInitial` is redefined:

```csharp
 1  void Main()
 2  {
 3     var userFaker = new Faker<User>()
 4     .RuleFor(u => u.FirstName, f => f.Name.FirstName())
 5     .RuleFor(u => u.LastName, f => f.Name.LastName())
 6     .RuleFor(u => u.MiddleInitial, (f, u) => u.FirstName[0]);
 7     
 8     // works
 9     userFaker.Generate().Dump();
10     
11     // Re-define the rule to be dependent on the LastName instead.
12     userFaker.RuleFor(u => u.MiddleInitial, (f, u) => u.LastName[0]);
13     // works
14     userFaker.Generate().Dump();
15  }
```

I hope that's clear. Please let me know if you have any other questions. Also, see #279 for more information if you need "Reordering" behavior (to move initially defined rules to the end of the execution list) instead of the default "Slot" behavior.

Thanks,
Brian
